frontend/simple/js/state.js

```diff
@@ -167,38 +170,36 @@ const actions = {
       //       and throw an exception and write a log message.
       return console.error(`NOT EXPECTING EVENT!`, contractId, entry)
     }
-
     const type = entry.type
-    entry = Events[type].fromObject(entry, hash)
-    if (entry instanceof Events.HashableContract) {
+    entry = Events[type] ? Events[type].fromObject(entry, hash) : contracts[type].fromObject(entry, hash)
+    if (Events.HashableContract.prototype.isPrototypeOf(entry)) {
       console.log(`handleEvent for new ${type}:`, entry)
       if (entry.toObject().parentHash) {
         // TODO: use a global notification object to handle this and all other errors
         return console.error(`${type} has non-null parentHash!`, entry)
       }
       await db.addLogEntry(contractId, entry)
       commit('addContract', {contractId, type, data: entry.toVuexState()})
-    } else if (entry instanceof Events.HashableAction) {
-      const contractType = state.contracts[contractId]
+    } else if (Events.HashableAction.prototype.isPrototypeOf(entry)) {
+      const contractType = state.contracts[contractId].type
       console.log(`handleEvent for ${type} on ${contractType}:`, entry)
-      if (!Events[contractType].isActionAllowed(state[contractType], entry)) {
+      const contract = Events[contractType] || contracts[contractType]
+      if (!contract.isActionAllowed(state[contractType], entry)) {
```

Comment:
This doesn't need to be changed from it was before, because actions are always in `Events`. You can revert and add a `// NOTE: ...` comment to that effect